File: ios/Shared/OutdoorGlanceContract/OutdoorGlanceSnapshot.swift
import Foundation

struct OutdoorGlanceSnapshot: Codable, Equatable, Sendable {
    static let currentSchemaVersion = 1

    let schemaVersion: Int
    let snapshotId: UUID
    let generatedAt: Date
    let freshness: OutdoorGlanceFreshness
    let trip: OutdoorGlanceTrip?
    let location: OutdoorGlanceLocation?
    let altitude: OutdoorGlanceAltitude?
    let weather: OutdoorGlanceWeather?
    let sun: OutdoorGlanceSun?
    let activity: OutdoorGlanceActivity?
    let parking: OutdoorGlanceParking?
    /// Language the user picked inside the iPhone app, so the watch can match it
    /// instead of following its own system locale. Optional and defaulted so
    /// snapshots written by older builds still decode at the same schema version.
    let language: String?
    /// Today's spend, already formatted in the user's home currency by the app.
    let todaySpendText: String?
    /// Decided on the iPhone, whose locale is reliable. The watch bundle keeps
    /// degrading Locale.current to its development language, which is how a
    /// user in Chengdu ended up reading altitude in feet.
    let usesMetric: Bool?

    init(
        schemaVersion: Int,
        snapshotId: UUID,
        generatedAt: Date,
        freshness: OutdoorGlanceFreshness,
        trip: OutdoorGlanceTrip?,
        location: OutdoorGlanceLocation?,
        altitude: OutdoorGlanceAltitude?,
        weather: OutdoorGlanceWeather?,
        sun: OutdoorGlanceSun?,
        activity: OutdoorGlanceActivity?,
        parking: OutdoorGlanceParking?,
        language: String? = nil,
        todaySpendText: String? = nil,
        usesMetric: Bool? = nil
    ) {
        self.schemaVersion = schemaVersion
        self.snapshotId = snapshotId
        self.generatedAt = generatedAt
        self.freshness = freshness
        self.trip = trip
        self.location = location
        self.altitude = altitude
        self.weather = weather
        self.sun = sun
        self.activity = activity
        self.parking = parking
        self.language = language
        self.todaySpendText = todaySpendText
        self.usesMetric = usesMetric
    }

    func isStale(at date: Date = Date()) -> Bool {
        date > freshness.validUntil
    }
}

struct OutdoorGlanceFreshness: Codable, Equatable, Sendable {
    let validUntil: Date
}

struct OutdoorGlanceTrip: Codable, Equatable, Sendable {
    let id: String
    let name: String
    let dayNumber: Int?
}

struct OutdoorGlanceLocation: Codable, Equatable, Sendable {
    let name: String?
    let latitude: Double
    let longitude: Double
    let horizontalAccuracyMeters: Double?
    let capturedAt: Date
}

struct OutdoorGlanceAltitude: Codable, Equatable, Sendable {
    let meters: Double
    let verticalAccuracyMeters: Double?
    let capturedAt: Date
}

struct OutdoorGlanceWeather: Codable, Equatable, Sendable {
    let temperatureCelsius: Double?
    let apparentTemperatureCelsius: Double?
    let conditionCode: String?
    let precipitationProbability: Double?
    let updatedAt: Date
}

struct OutdoorGlanceSun: Codable, Equatable, Sendable {
    let sunriseAt: Date?
    let sunsetAt: Date?
}

struct OutdoorGlanceActivity: Codable, Equatable, Sendable {
    let steps: Int?
    let distanceMeters: Double?
    let updatedAt: Date
}

struct OutdoorGlanceParking: Codable, Equatable, Sendable {
    let savedAt: Date
    let latitude: Double
    let longitude: Double
    let distanceMeters: Double?
    let bearingDegrees: Double?
}

enum OutdoorGlanceContractError: Error, Equatable {
    case unsupportedSchemaVersion(Int)
}

enum OutdoorGlanceTransport {
    static let snapshotDataKey = "outdoorGlanceSnapshotV1"
}

enum OutdoorGlanceCodec {
    static func encode(_ snapshot: OutdoorGlanceSnapshot) throws -> Data {
        try makeEncoder().encode(snapshot)
    }

    static func decode(_ data: Data) throws -> OutdoorGlanceSnapshot {
        let snapshot = try makeDecoder().decode(
            OutdoorGlanceSnapshot.self,
            from: data
        )

        guard snapshot.schemaVersion == OutdoorGlanceSnapshot.currentSchemaVersion else {
            throw OutdoorGlanceContractError.unsupportedSchemaVersion(
                snapshot.schemaVersion
            )
        }

        return snapshot
    }

    private static func makeEncoder() -> JSONEncoder {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]

        encoder.dateEncodingStrategy = .custom { date, encoder in
            var container = encoder.singleValueContainer()
            try container.encode(makeFormatter(fractionalSeconds: true).string(from: date))
        }

        return encoder
    }

    private static func makeDecoder() -> JSONDecoder {
        let decoder = JSONDecoder()

        decoder.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let value = try container.decode(String.self)

            if let date = makeFormatter(fractionalSeconds: true).date(from: value) {
                return date
            }

            if let date = makeFormatter(fractionalSeconds: false).date(from: value) {
                return date
            }

            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "Invalid ISO 8601 date: \(value)"
            )
        }

        return decoder
    }

    private static func makeFormatter(
        fractionalSeconds: Bool
    ) -> ISO8601DateFormatter {
        let formatter = ISO8601DateFormatter()

        formatter.formatOptions = fractionalSeconds
            ? [.withInternetDateTime, .withFractionalSeconds]
            : [.withInternetDateTime]

        return formatter
    }
}

File: lib/watch/OutdoorGlanceSnapshot.ts
export const OUTDOOR_GLANCE_SNAPSHOT_SCHEMA_VERSION = 1 as const;
export const OUTDOOR_GLANCE_DEFAULT_TTL_MS = 30 * 60 * 1000;

export type ISODateString = string;

export type OutdoorGlanceSnapshot = {
  schemaVersion: typeof OUTDOOR_GLANCE_SNAPSHOT_SCHEMA_VERSION;
  snapshotId: string;
  generatedAt: ISODateString;
  freshness: OutdoorGlanceFreshness;
  trip: OutdoorGlanceTrip | null;
  location: OutdoorGlanceLocation | null;
  altitude: OutdoorGlanceAltitude | null;
  weather: OutdoorGlanceWeather | null;
  sun: OutdoorGlanceSun | null;
  activity: OutdoorGlanceActivity | null;
  parking: OutdoorGlanceParking | null;
  /** Language selected inside the iPhone app, so the watch can match it. */
  language: string | null;
  /** Today's spend, already formatted in the home currency. */
  todaySpendText: string | null;
  /** Whether the user's region is metric, decided on the phone. */
  usesMetric: boolean | null;
};

export type OutdoorGlanceFreshness = {
  validUntil: ISODateString;
};

export type OutdoorGlanceTrip = {
  id: string;
  name: string;
  dayNumber: number | null;
};

export type OutdoorGlanceLocation = {
  name: string | null;
  latitude: number;
  longitude: number;
  horizontalAccuracyMeters: number | null;
  capturedAt: ISODateString;
};

export type OutdoorGlanceAltitude = {
  meters: number;
  verticalAccuracyMeters: number | null;
  capturedAt: ISODateString;
};

export type OutdoorGlanceWeather = {
  temperatureCelsius: number | null;
  apparentTemperatureCelsius: number | null;
  conditionCode: string | null;
  precipitationProbability: number | null;
  updatedAt: ISODateString;
};

export type OutdoorGlanceSun = {
  sunriseAt: ISODateString | null;
  sunsetAt: ISODateString | null;
};

export type OutdoorGlanceActivity = {
  steps: number | null;
  distanceMeters: number | null;
  updatedAt: ISODateString;
};

export type OutdoorGlanceParking = {
  savedAt: ISODateString;
  latitude: number;
  longitude: number;
  distanceMeters: number | null;
  bearingDegrees: number | null;
};

export type DateLike = Date | string | number;

export type OutdoorGlanceTripDraft = {
  id: string | number;
  name: string;
  dayNumber?: number | null;
};

export type OutdoorGlanceLocationDraft = {
  name?: string | null;
  latitude: number | string;
  longitude: number | string;
  horizontalAccuracyMeters?: number | string | null;
  capturedAt?: DateLike | null;
};

export type OutdoorGlanceAltitudeDraft = {
  meters: number | string;
  verticalAccuracyMeters?: number | string | null;
  capturedAt?: DateLike | null;
};

export type OutdoorGlanceWeatherDraft = {
  temperatureCelsius?: number | string | null;
  apparentTemperatureCelsius?: number | string | null;
  conditionCode?: string | number | null;
  precipitationProbability?: number | string | null;
  updatedAt?: DateLike | null;
};

export type OutdoorGlanceSunDraft = {
  sunriseAt?: DateLike | null;
  sunsetAt?: DateLike | null;
};

export type OutdoorGlanceActivityDraft = {
  steps?: number | string | null;
  distanceMeters?: number | string | null;
  updatedAt?: DateLike | null;
};

export type OutdoorGlanceParkingDraft = {
  savedAt?: DateLike | null;
  latitude: number | string;
  longitude: number | string;
  distanceMeters?: number | string | null;
  bearingDegrees?: number | string | null;
};

export type ComposeOutdoorGlanceSnapshotInput = {
  now?: DateLike;
  ttlMs?: number;
  trip?: OutdoorGlanceTripDraft | null;
  location?: OutdoorGlanceLocationDraft | null;
  altitude?: OutdoorGlanceAltitudeDraft | null;
  weather?: OutdoorGlanceWeatherDraft | null;
  sun?: OutdoorGlanceSunDraft | null;
  activity?: OutdoorGlanceActivityDraft | null;
  parking?: OutdoorGlanceParkingDraft | null;
  language?: string | null;
  todaySpendText?: string | null;
  usesMetric?: boolean | null;
};
